### ISSUE TITLE
Prometheus query issue : kayenta builds same query for canary & baseline : relates to #540

### ISSUE THREAD

**Issue by amsin02** [spinnaker/kayenta#553]
Hey,
I tried using Prometheus as me METRIC_STORE. But, Kayenta makes the same query for canary and baseline even after supplying all the parameters.

**kayenta.yml**
```
  prometheus:
    enabled: true
    accounts:
      - name: prometheus-account
        endpoint:
          baseUrl: http://localhost:9090
        supportedTypes:
          - METRICS_STORE
        defaultScopeKey: server_group
        defualtLocationKey: server_region
```
===============================================

**Execution Request**

```
 "executionRequest": {
    "analysisIntervalMins": 2,
    "beginAfterMins": 0,
    "lifetimeDurationMins": 10,
    "lookbackMins":0,
    "scopes": [
      {
        "controlScope": "baseline",
        "experimentScope": "canary",
        "scopeName": "default",
        "step": 10
      }
    ],
    "siteLocal": {},
    "thresholds": {
      "marginal": 95.0,
      "pass": 75.0
    }
  }
```
================================================

I checked the **metricSetListId** which I got in the response to check what queries it is making to Prometheus.

**MetricSetListQueries**
```
"attributes": {
      "control": {
        "query": "rate(num_of_error_requests[1m])"
      },
      "experiment": {
        "query": "rate(num_of_error_requests[1m])"
      }
```
===================================================

whereas for **SignalFX**, it supports different queries for canary & baseline for the same execution request:

**MetricSetListQueries**
```
"attributes":{ 
"control": { 
 "signal-flow-program": "data('num_of_success_requests', filter=filter('server_group','baseline')).sum(by=['server_group']).publish()" },

 "experiment": {
 "signal-flow-program" :  "data('num_of_success_requests', filter=filter('server_group', 'canary')).sum(by=['server_group']).publish()"}
}
```

**Comment by fieldju** [comment]
@amsin02 

I was just talking to some people about this in #kayenta on the Spinnaker Slack server.

I am not super familiar with the Prometheus Kayenta integration and how it uses scope and location to build queries.

I would recommend checking out the [unit tests for the query building logic](https://github.com/spinnaker/kayenta/blob/master/kayenta-prometheus/src/test/groovy/com/netflix/kayenta/prometheus/metrics/PrometheusMetricsServiceSpec.groovy#L71-L90).

There seems to be 3 patterns for defining how your queries get built and I would need to reverse engineer things starting from that unit test to educate myself how it all works.

Here some links of interest I found

- https://www.mirantis.com/blog/spinnaker-canary-pipelines-how-to-set-up-kayenta-with-prometheus/
- http://docs.opsmx.com/additional-feature-configuration/monitoring-tools/prometheus-kubernetes/
- https://www.spinnaker.io/guides/user/canary/config/#add-filter-templates 

From what I can gather is that you need to use some sort of template functionality to get Prometheus to generate queries that are capable of collecting the data for the Baseline and Experiment separately.

**Comment by rarneson** [comment]
@amsin02 I'm working through the same thing right now. I'm actually using `customInlineTemplate` to write my PromQL, along the lines of what @fieldju  lined to above, specifically https://github.com/spinnaker/kayenta/blob/master/kayenta-prometheus/src/test/groovy/com/netflix/kayenta/prometheus/metrics/PrometheusMetricsServiceSpec.groovy#L89.

I don't quite have it working yet, but I'll post what I currently have in the hopes it gives you some ideas. 

I'm using the `/canary` endpoint and passing in a `canaryAdHocExecutionRequest` object which contains both my `canaryConfig` and `executionRequest`.

```json
{
  "canaryConfig": {
    "name": "app-config",
    "description": "App Canary Configuration",
    "configVersion": "1",
    "applications": [
      "app"
    ],
    "judge": {
      "name": "NetflixACAJudge-v1.0",
      "judgeConfigurations": {}
    },
    "metrics": [
      {
        "name": "404 Status Codes",
        "query": {
          "customInlineTemplate": "PromQL:app_request_status{code=\"404\", app=\"${location}\"}",
          "serviceType": "prometheus",
          "type": "prometheus",
          "scopeName": "default"
        },
        "groups": [
          "StatusCodes"
        ],
        "analysisConfigurations": {
          "canary": {
            "direction": "increase"
          }
        },
        "scopeName": "default"
      }
    ],
    "templates": {},
    "classifier": {
      "groupWeights": {
        "StatusCodes": 100.0
      },
      "scoreThresholds": {
        "marginal": 75.0,
        "pass": 75.0
      }
    }
  },
  "executionRequest": {
    "scopes": {
      "default": {
        "controlScope": {
          "scope": "control",
          "location": "app-prod",
          "start": "2019-06-07T07:00:00.000Z",
          "end": "2019-06-07T09:00:00.000Z",
          "step": 2
        },
        "experimentScope": {
          "scope": "experiment",
          "location": "app-canary",
          "start": "2019-06-07T07:00:00.000Z",
          "end": "2019-06-07T09:00:00.000Z",
          "step": 2
        }
      }
    },
    "siteLocal": {},
    "thresholds": {
      "pass": 95,
      "marginal": 75
    }
  }
}
```

What I'm expecting is that the PromQL in my config object can reference the `location` keys in the `controlScope` and `experimentScope` from the `executionRequest` object by `${location}`. 

Unfortunately it results in this in the response:
```json
"exception": {
    "details": {
      "error": "Failed Expression Evaluation",
      "errors": [
        "Failed to evaluate [customInlineTemplate] : PromQL:adaptiveui_request_status{code=\"404\", app=\"location\"} not found"
      ]
    }
  }
```

I'll update here if/when I get it writing the correct queries for control and experiment. Hope this helps.

**Comment by amsin02** [comment]
Hi @fieldju,
I already followed the links you mentioned to correctly configure Prometheus for Kayenta, still no luck.

======================================================

@rarneson, yeah I have tried doing the same with **customFilterTemplate** 

``` "customFilterTemplate":"my_template"  ```

and setting template expression as :

``` "templates": {"my_template":"server_group=${scope}"} ```

I tried few other expressions also like
```
"templates": {"my_template":"scope=${server_group}"}
"templates": {"my_template":"scope=${server_region}"}
```
But. nothing seems to produce the expected results. Even though, I am publishing both **server_group** and **server_region** in my metrics. But, I get exceptions like

```
 "exception": {
          "details": {
            "errors": [
              null,
              "Failed to evaluate [my_template] : server_group=scope not found"
            ]
          }
```

**Comment by duftler** [comment]
Should be resolved by: https://github.com/spinnaker/kayenta/pull/579

**Comment by maggieneterval** [comment]
We just released [1.14.8](https://gist.github.com/spinnaker-release/f223e8c3790e2ded5cfde046965e658d), which includes @duftler's fix. Big thanks to Matt for fixing and everyone on this issue for helping to report / debug!
